taichi/rhi/window_system.cpp

```diff
@@ -0,0 +1,60 @@
+#include "window_system.h"
+#include "taichi/rhi/impl_support.h"
+
+#include <mutex>
+#include <array>
+#include <iostream>
+
+namespace taichi::lang::window_system {
+
+#ifdef TI_WITH_GLFW
+struct GLFWState {
+  std::mutex mutex;
+  int glfw_ref_count = 0;
+};
+
+static GLFWState glfw_state;
+
+static void glfw_error_callback(int code, const char *description) {
+  std::array<char, 1024> buf;
+  snprintf(buf.data(), buf.size(), "GLFW Error %d: %s", code, description);
+  RHI_LOG_ERROR(buf.data());
+}
+
+bool glfw_context_acquire() {
+  std::lock_guard lg(glfw_state.mutex);
+  if (glfw_state.glfw_ref_count == 0) {
+    auto res = glfwInit();
```

Comment:
How expensive are `glfwInit()` and `glfwTerminate()`. If they are costly, put these calls outside the mutex?

Reply:
Because they are global states, they need to be serial